lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+
+import argparse
+import asyncio
+import functools
+import logging
+import re
+import signal
+import subprocess
+import sys
+import textwrap
+from pathlib import Path
+from types import MappingProxyType
+from typing import List, Union
+
+from dbus_fast.aio import MessageBus
+from dbus_fast.constants import BusType, MessageType
+from dbus_fast.errors import DBusError
+from dbus_fast.message import Message
+from ops.charm import CharmBase
+from ops.framework import EventBase
+
+LIBID = "HPCTEAMUSEONLY"
+LIBAPI = 0
+LIBPATCH = 1
+PYDEPS = ["dbus-fast>=1.90.2"]
+
+_logger = logging.getLogger(__name__)
+_juju_unit = None
+_service_states = {}
+_service_hook_regex_filter = re.compile(r"service-(?P<service>[\w\\:-]*)-(?:started|stopped)")
+_DBUS_CHAR_MAPPINGS = MappingProxyType(
+    {
+        "_5f": "_",
+        "_40": "@",
+        "_2e": ".",
+        "_2d": "-",
+        "_5c": "\\",
+    }
+)
+
+
+def _systemctl(*args) -> None:
+    """Control systemd by via executed `systemctl ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if systemctl command fails.
+    """
+    cmd = ["systemctl", *args]
+    _logger.debug(f"systemd: Executing command {cmd}")
+    try:
+        subprocess.check_output(cmd)
+    except subprocess.CalledProcessError as e:
+        _logger.error(f"systemctl command failed: {e}")
+        raise
+
+
+_daemon_reload = functools.partial(_systemctl, "daemon-reload")
+_daemon_reload.__doc__ = "Reload systemd manager configuration."
+_start_service = functools.partial(_systemctl, "start")
+_start_service.__doc__ = "Start systemd service unit."
+_stop_service = functools.partial(_systemctl, "stop")
+_stop_service.__doc__ = "Stop systemd service unit."
+_enable_service = functools.partial(_systemctl, "enable")
+_enable_service.__doc__ = "Enable systemd service."
+_disable_service = functools.partial(_systemctl, "disable")
+_disable_service.__doc__ = "Disable systemd service."
+
+
+class ServiceStartedEvent(EventBase):
+    """Event emitted when service has started."""
+
+
+class ServiceStoppedEvent(EventBase):
+    """Event emitted when service has stopped."""
+
+
+class SystemdNotices:
+    """Observe systemd services on your machine base."""
+
+    def __init__(self, charm: CharmBase, services: Union[str, List[str]]) -> None:
+        """Instantiate systemd notices service."""
+        self._charm = charm
+        self._services = [services] if isinstance(services, str) else services
+        unit_name = self._charm.unit.name.replace("/", "-")
+        self._service_file = Path(f"/etc/systemd/system/juju-{unit_name}-systemd-notices.service")
+
+        _logger.debug(f"Attaching systemd notice events to charm {self._charm.__class__.__name__}")
+        for service in self._services:
+            self._charm.on.define_event(f"service_{service}_started", ServiceStartedEvent)
+            self._charm.on.define_event(f"service_{service}_stopped", ServiceStoppedEvent)
+
+    def subscribe(self) -> None:
```

Comment:
Not sure, but should this be called `start` to parallel `stop`?

Reply:
I named this method `subscribe(...)` rather than `start(...)` because it does more than just start the `juju-systemd-notices` daemon in the background. It:

1. Creates the hooks for the service start/stop events
2. Creates the `juju-systemd-notices` service file
3. Reloads the systemd manager configuration
4. Enables the daemon (so it will start even after machine reboot)
5. Starts the `juju-systemd-notices` daemon

I debated naming the method `start(...)`, but given that it does more than just start the daemon, I didn't want to create any confusion. E.g. why is this method creating a service/creating hooks off of the dispatch hook? Let me know what your thoughts are about this.

Reply:
Yeah, makes sense. I'd be fine with `subscribe`. Could also consider `start_and_subscribe` -- a bit of a mouthful, but very clear.